galaxyproject/galaxy
# A replacement for manage_db.sh + rewrite of migration system documentation

This is a replacement for the `manage_db.sh` script + new migration documentation.

**Rationale**
The pre-22.05 `manage_db.sh` was a thin wrapper around the SQLAlchemy Migrate CLI runner. With the move to Alembic, this script was rewritten to be an adapter between the old interface supporting 2 essential commands (upgrade/downgrade) and the Alembic CLI. A new script, `run_alembic.sh` was added as a wrapper around Alembic's CLI, to provide access to the full scope of commands provided by Alembic.

However, the `manage_db.sh` script was very limited in what it could do, whereas `run_alembic.sh` required more verbose command syntax, familiarity with advanced Alembic features + an understanding of how Galaxy's data model was implemented (the separation of galaxy and tool shed install models, implemented with Alembic branches).

There was much discussion about designing a better solution that might have these characteristics:
- tool-agnostic interface (not dependent on Alembic or SQLAlchemy Migrate)
- hide the complexity of Alembic branches and the setup of  Galaxy's data model
- provide a set of commands with simple and consistent syntax that enable commonly used functionality. 

This PR proposes a solution.

**Implementation notes**
The script is implemented with argparse (so it provides proper command line help). It supports a set of Alembic CLI commands and command options that may be useful in the context of developing or administering Galaxy. 

The tool targets only the *gxy* data model/migration branch (with a few exceptions, such as running `./db.sh upgrade` without specifying a revision, which is explained in the documentation).

The documentation has been completely rewritten. It provides a better overview of the migration system, detailed descriptions of available commands and options, Galaxy-specific usage examples, etc.

[EDIT: reverted to manage_db.sh after code review.]
~~Why rename `manage_db.sh` to `db.sh`?~~ The new script comes with new commands: a new name makes it explicit, whereas the old name might cause confusion (what version of `manage_db.sh` am I using?). Also, I think, `db.sh` is more succinct: `db.sh upgrade` and `db.sh revision` reads as "database upgrade" and "database revision", whereas `manage_db sh [operation]` is a little less obvious: "manage database upgrade". But that's completely subjective, and I have no objections to keeping `manage_db.sh` if that's a better approach.

The old `manage_db.sh` script translated command line arguments into a format required by the Alembic CLI runner. This new script does not invoke the Alembic script; instead, it invokes Alembic commands programmatically via the Alembic API. As a result, a lot of our code became unnecessary - so it has been deleted. Also, I think the new implementation is more straightforward and less brittle.

**Tests**
Overall approach:
1. Use a blank test database, store revision scripts in a temporary directory, leave the rest unchanged.
2. Use Alembic API for test setup and for accessing the database.
3. Run the `db.sh` command as a subprocess, verify captured output + database state.

In other words, there's almost no additional infrastructure involved: we use Alembic for any setup, and Python's subprocess module for executing the script. 

The tests run against SQLite only. We can run them against Postgres as well, but, I think, that's not necessary: the tests for this script verify that Alembic commands are invoked correctly and produce the expected results; we don't need to test how Alembic interacts with another database.

Here're the details of setting up the staging environment:
- Create staging location (`/tmp`)
- Create test database (SQLite in `/tmp`)
- Copy Galaxy's `alembic.ini` to staging location, overwriting:
  - `sqlalchemy.url` (url of test database)
  - `version_locations` (staging location)
  - `script_location` (point to `lib/galaxy/model/migrations/alembic/` - which is where the real `env.py` is located)
- Create *gxy* and *tsi* branches

**Next steps (separate PRs):**
- [ ] Handle cases when database (one or both) is in pre-22.05 state (container-based solution) + update documentation
- [ ] Same as above, for downgrading to pre-22.05 state + update documentation
- [ ] Update relevant tutorials and slides on GTN

## How to test the changes?
(Select all options that apply)
  1. Run `./db.sh -h` to view the help and/or `./db.sh [command] -h` to view help for individual commands.
  2. Try running each command. You may use the `-c` option to specify an alternate galaxy.yml file not to modify your database(s).

## License
